### PR TITLE
Update Seed Generator

### LINKED ISSUE #650: current rng setup is full of footguns in jax
right now unseeded calls to e.g. `keras.random.uniform` are going to acquire static seeds at trace time. this has a few undesirable consequences:

1) subsequent calls will have the same randomness each time (e.g. dropout will have a fixed mask instead of random each step)
2) the jax compiler cache will ~never hit, as the constant rng seed values will be different every time

to get around this, some kind of rng state management is necessary. flax does this with hierarchical management of rng's from the `Scope`. such an approach is fairly complex however, and there might be simpler options e.g. a single global `rng` state, which gets included with the training state in `model.fit`, unseeded rng calls would then do something along the lines of

```
state.seed, local_seed = jax.random.split(state.seed)
```

### PR DESCRIPTION
**Pull request by AakashKumarNain**
This PR is an attempt to fix #650 A few notable things:

1. Right now, I have kept `42` as the default seed so that everyone can test it easily on their side with the same setup. Will replace it with a random number once PR is approved
2. `torch` is a pain to work with when it comes to random generation. Why? Because it consumes and updates the state of the generator implicitly. Had it followed the same algo as TF and JAX, we could have avoided a couple of `if-else` conditions
3. I have tested the workflow on my end with all the backends with random ops, and layers. It works flawlessly! I will update the tests later on.

Here is an overview of how the seed generation works now:

### Scenario 1: No seed value is provided

```python
samples = ops.random.randint(shape=(1,), minval=0, maxval=10)
```
1. Once this random op is called, it will call the `draw_seed(...)` function.
2. It will check if we already have a global seed generator. If yes, then we will use call the `next(...)` method to get the next state of the global generator and draw samples from this updated rng state
3. If there is no global seed generator yet, It will create an instance of `SeedGenerator()` with `seed=None`, which will initialize the seed state with `42`, and will set the global seed generator to this instance
4. Subsequent calls of unseeded random ops will use the next state of this global seed generator

### Scenario 2: seed value is provided

```python
samples = ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)
```
~1. Once this random op is called, it will call the `draw_seed(...)` function.~
~2. If the seed is an `int`, it will create a `SeedGenerator` instance by calling the `make_initial_seed(...)` function.~
~3. Or you can instantiate `SeedGenerator` and pass to it.~
~4. The function then will use the seed value produced by calling the `next(..)` method in `draw_seed(...)` function~

1. Once this random op is called, it will call the `draw_seed(...)` function.
2. The `draw_seed()` function will call the `make_initial_seed(seed)` function that expects an integer argument. Only integer arguments are supported to make the behavior consistent across backends. 
3. This is a `stateless` call, meaning if you call the function with the same seed, you will get the same results everytime.

---

### Scenario 1: No seed value is provided

1. It will check if a global seed generator is there. If yes, will draw seed by calling `next(...)`
2. If no global seed generator is present, it will create one with `42` as the initial state. And it will then draw seed from this global seed generator on subsequent calls.

### Scenario 2: Seed value is provided

Will instantiate a `SeedGenerator`, and subsequent calls will draw seed from this generator.

**Note:** The `SeedGenerator` class now depends on the backend specific state updates of the rng. In case of `TF` and `JAX` we will split the state at every call while in torch the state is consumed implicitly. The `next()` method in the `SeedGenerator` is now fully deterministic and depends on the algos used by the backend for managing state of the rng

Though I don't think I have missed anything in testing this on my end, it will be good if someone can confirm the same on their end. cc: @fchollet @martin-gorner @GallagherCommaJack

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR!

> Scenario 2: Seed value is provided

This is not intended behavior -- if you do `ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)` you should get the same value at every call. It should not be stateful. It would be very counter-intuitive if it were stateful and the same seed gave you different results at every call.

**Comment by AakashKumarNain** [comment]
> Thanks for the PR!
> 
> > Scenario 2: Seed value is provided
> 
> This is not intended behavior -- if you do `ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)` you should get the same value at every call. It should not be stateful. It would be very counter-intuitive if it were stateful and the same seed gave you different results at every call.

@fchollet Sorry for the wrong wording here. It does exactly what you expect. On every call `ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)`  will give you the same result. Here is an example to show the same:

```python
for _ in range(5):
    samples = backend.random.randint(shape=(1,), minval=0, maxval=10, seed=1)
    print(samples)

# output
[4]
[4]
[4]
[4]
[4]
```

**Comment by fchollet** [comment]
> You will get different values only if are making subsequent calls to the same function

This is what I mean. The behavior of your code should not change if you put it in a traced function. In this case, `ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)` should return the same value regardless of the context, including in the case where it's a traced function that is called multiple times. Basically, we should keep the current RNG seed behavior.

**Comment by AakashKumarNain** [comment]
Okay, I will refactor the code. But clearing it again:

`ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)`
The above will return the same result now matter how many times you call it

`ops.random.randint(shape=(1,), minval=0, maxval=10, seed=SeedGenerator(1))`
The above will return the same result now matter how many times you call it


Re: The remaining case where we are passing the **update rng state** to a seeded function like this

```python
s = SeedGenerator(1)
for _ in range(3):
    samples = ops.random.randint(shape=(1,), minval=0, maxval=10, seed=s)
    print(samples)
    
# output
[4]
[0]
[1]
```

This behavior is consistent with the below workflow

```python
gen = torch.Generator().manual_seed(1)
for _ in range(3):
    samples = torch.randint(low=0, high=10, size=(1,), generator=gen)
    print(samples)

# output
tensor([5])
tensor([9])
tensor([4])

###################

gen = tf.random.Generator.from_seed(1)
seed = tf.cast(gen.make_seeds(2)[0], tf.int32)
for _ in range(3):
    seed, subseed = tf.random.split(seed, 2)
    samples = tf.random.stateless_uniform(minval=0, maxval=10, shape=(1,), dtype=tf.int32, seed=seed)
    print(samples)

# output
tf.Tensor([0], shape=(1,), dtype=int32)
tf.Tensor([5], shape=(1,), dtype=int32)
tf.Tensor([4], shape=(1,), dtype=int32)

########################

key = jax.random.PRNGKey(1)
for _ in range(3):
    key, subkey = jax.random.split(key)
    samples = jax.random.randint(key, minval=0, maxval=10, shape=(1,))
    print(samples)

# output
[5]
[9]
[1]
```

**Comment by AakashKumarNain** [comment]
@fchollet I have made the changes. Random functions are now stateless while layers depending on the rng state are stateful and uses `SeedGenerator`. I have edited the note at the top for the details. Please let me know if you want to change anything else. At this point

`ops.random.randint(shape=(1,), minval=0, maxval=10, seed=1)`
This will always return the same result regardless of the context, and 

`ops.random.randint(shape=(1,), minval=0, maxval=10, seed=None)`
The results here will depend on the state of the global random generator

File: keras_core/backend/jax/random.py [hunk 1]
```diff
@@ -3,16 +3,35 @@
...
+            f"Initial seed should be a scalar value. Received seed={seed}."
+        )
+    if seed < 0:
+        raise ValueError(
+            f"Seed should be a non-negative number. Received seed={seed}."
+        )
+    return None, jax.random.PRNGKey(seed=seed)
+
+
+def get_next_state(seed):
```
Comment (fchollet) [inline keras_core/backend/jax/random.py:31]: This function should have "seed" somewhere in the name since it's part of the same group of functions, e.g. `get_next_seed_state()`

File: keras_core/backend/jax/random.py [hunk 2]
```diff
@@ -3,16 +3,35 @@
...
-    if isinstance(seed, jax.Array):
-        return seed
+    if isinstance(seed, SeedGenerator):
+        return seed.next()
     else:
         return draw_seed(seed)
 
 
+def make_default_seed():
+    return None, jax.random.PRNGKey(42)
```
Comment (fchollet) [inline keras_core/backend/jax/random.py:16]: The default seed should not be fixed, it should be drawn from Python `random`

File: keras_core/backend/tensorflow/random.py [hunk 3]
```diff
@@ -4,12 +4,37 @@
...
-    # TF ops only accept int32/64 seeds but our base seed is uint32.
-    return tf.cast(draw_seed(seed), dtype="int32")
+    if isinstance(seed, SeedGenerator):
+        return seed.next()
+    else:
+        return draw_seed(seed)
+
+
+def make_default_seed():
+    rng = tf.random.Generator.from_seed(42)
```
Comment (fchollet) [inline keras_core/backend/tensorflow/random.py:17]: Same here

File: keras_core/random/seed_generator.py [hunk 4]
```diff
@@ -52,49 +48,56 @@ def __init__(self, seed=None, **kwargs):
...
-                "Argument `seed` must be an integer. " f"Received: seed={seed}"
-            )
+            rng, seed = backend.random.make_default_seed()
+        else:
+            rng, seed = backend.random.make_initial_seed(seed)
 
-        def seed_initializer(*args, **kwargs):
-            dtype = kwargs.get("dtype", None)
-            return self.backend.convert_to_tensor([seed, 0], dtype=dtype)
+        if backend.backend() == "tensorflow":
```
Comment (fchollet) [inline keras_core/random/seed_generator.py:56]: Why is this needed? Can't we use `uint32` everywhere?

File: keras_core/random/seed_generator.py [hunk 5]
```diff
@@ -52,49 +48,56 @@ def __init__(self, seed=None, **kwargs):
...
-        seed_state = self.state
-        # Use * 1 to create a copy
-        new_seed_value = seed_state.value * 1
-        if ordered:
-            increment = self.backend.convert_to_tensor(
-                np.array([0, 1]), dtype="uint32"
-            )
-            self.state.assign(seed_state + increment)
+    def next(self):
+        if backend.backend() == "torch":
```
Comment (fchollet) [inline keras_core/random/seed_generator.py:72]: Can we move the torch specific code to get_next_seed_state?

File: keras_core/random/seed_generator.py [hunk 6]
```diff
@@ -103,49 +106,9 @@ def global_rng_state():
...
-    seed = jax.numpy.array([0, 0])
-    x, y, seed = random_numbers(seed)
-    new_x, new_y, seed = random_numbers(seed)
-    ```
     """
-    return global_seed_generator().state
 
-
-def make_default_seed():
-    return python_random.randint(1, int(1e9))
```
Comment (fchollet) [inline keras_core/random/seed_generator.py:135]: This is the code we should use for the default seed instead of `42`

File: keras_core/backend/jax/random.py (hunk 1 above)
Comment (AakashKumarNain) [reply keras_core/backend/jax/random.py:31]: Will rename it.

File: keras_core/backend/jax/random.py (hunk 2 above)
Comment (AakashKumarNain) [reply keras_core/backend/jax/random.py:16]: I know. I kept if just for the sake of demonstration. Will replace it with a random selection

File: keras_core/backend/tensorflow/random.py (hunk 3 above)
Comment (AakashKumarNain) [reply keras_core/backend/tensorflow/random.py:17]: Same as above

File: keras_core/random/seed_generator.py (hunk 4 above)
Comment (AakashKumarNain) [reply keras_core/random/seed_generator.py:56]: TF random generator use int32/int64, and with XLA only int32 is allowed. Casting it to uint32 can cause some unintended consequences, and will be hard to debug

File: keras_core/random/seed_generator.py (hunk 5 above)
Comment (AakashKumarNain) [reply keras_core/random/seed_generator.py:72]: Let me elaborate why is it coded this way. 
1. In TF/JAX, when we get the next state by splitting the seed, we get two states. We assign one of these states to the state of the `SeedGenerator`, and we also pass the same state to the random functions
2. In torch, we don't have a way to control the random state explicitly. We keep passing the same gen object to random functions, and they consume and update the state of the gen implicitly. If we move this to torch backend, even then we would be required to have this condition

File: keras_core/random/seed_generator.py (hunk 6 above)
Comment (AakashKumarNain) [reply keras_core/random/seed_generator.py:135]: Yes, I am aware of this. I will make that change in the final pass when I rewrite tests

**Comment by AakashKumarNain** [comment]
Even though I have made changes to the seed generator class, this won't work for jit compiled TF because compiled TF function supports only singleton variable creation.

Similarly there is a weird issue on JAX side where it keeps complaining about the dtype of PRNGKey `TypeError: JAX encountered invalid PRNG key data: expected key_data.dtype = uint32; got dtype=uint32`

Our only options are:
1. Either keep the current SeedGenerator class as it is. But for the JAX users, the current design doesn't guarantee the reproducibility that comes with JAX PRNG design (unless we have some proof for it).
2. The other option is to have a SeedGenerator class for each backend. That we, we can over come a lot of these issues

**Comment by fchollet** [comment]
> Also, for JAX backend, we need to raise an issue whenever seed=None is passed instead of using a global state

This is too general -- I think specifically we should detect when the global seed generator is being used in a tracing context in JAX, and throw an error at that time. It is fine to use `seed=None` in JAX when running eagerly.

> The other option is to have a SeedGenerator class for each backend. That we, we can over come a lot of these issues.

If we are disabling tracing + `seed=None` in JAX, what are the remaining benefits of having backend specific generators? Just the seed-splitting algo? I am fundamentally not worried about it. The current setup is reproducible and works well. If you are in a multi worker setting, you know that the seed on the worker deterministically depends on the Python RNG seed state, so you can chose to see `random` with the same seed on each worker if you want the same random numbers getting generator, or seed each worker with e.g. `worker_id` to get different random behavior on each worker.

**Comment by AakashKumarNain** [comment]
>  It is fine to use seed=None in JAX when running eagerly.

There are two **core** issues with this approach:
1. It diverges from the behvior of the backend. If no seed is passed, we should raise the same error as raised byJAX irrespective of the execution mode. 
2. It changes the mental model for the backend. The idea of passing a seed in  random functions in JAX, and that too not a global seed generator, is about controlling the entropy explicitly. If we don't mimic the same behavior, we then become a "specailzed" implementation of JAX backend.

> what are the remaining benefits of having backend specific generators?

Listing a few of them:

* The PRNG design in JAX makes it easy to escape the sequential execution constraint, making it parallelizable and reproducible at the same time. To illustrate, let's take a simple example:

```python

def random_letter():
    return random.choice(["a", "b", "c"])

def random_number():
   return random.randint(0, 10, shape=())

res = random_letter() + str(random_number())
```

Both the random functions in the above example are independent of each other, meaning that we should be able to run them in any order. We don't know which function will be executed by the compiler first. If a global state is used, the result will depend on which function was executed first making it completely irreproducible.  Even if we use a stateful seed here, it introduces another sequential dependency. **Both of these things don't align with the XLA semantics used by JAX.**

* Though you can use any counter based PRNG algo, but the only way to ensure that there is no sequential threading, and that we are aligned with the functional programming is to have a **splittable** PRNG instead of a linear one. [Here](https://dl.acm.org/doi/10.1145/2503778.2503784) is the paper for the same.
